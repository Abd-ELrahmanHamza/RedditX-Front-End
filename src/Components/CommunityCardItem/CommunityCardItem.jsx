// Imports
import { useState } from "react";
import { BASE_URL } from "API/axios";
import {
  Caret,
  CommunityNameChild,
  CommunityNameContainer,
  Container, HeaderLine,
  Image,
  Item,
  JoinBtn,
  JoinContainer,
  Number,
} from "./CommunityCardItem.styled";
import { Link } from "react-router-dom";
import { FaAngleUp, FaAngleDown } from "react-icons/fa";

/**
 * Component that  shows the names of communities up-to-date.
 * @returns {Component.React}
 */
const CommunityCardItem = ({ communityId, community, srIcon }) => {
  const [btnContent, setBtnContent] = useState("Join");

  const clickHandler = (e) => {
    e.preventDefault();
    if (btnContent === "Join") {
      setBtnContent("Joined");
    } else {
      setBtnContent("Join");
    }
  };

  const mouseEnterHandler = () => {
    if (btnContent === "Joined") {
      setBtnContent("Leave");
    }
  };

  const MouseLeaveHandler = () => {
    if (btnContent === "Leave") {
      setBtnContent("Joined");
    }
  };

  return (
      <>
    <Container>
      <Link to={"/subreddit/*"}>
        <Item>
          <Number>{communityId}</Number>
          <Caret>
            <FaAngleUp />
          </Caret>
          <Image
            crossOrigin="anonymous"
            src={`${BASE_URL}/subreddits/files/${srIcon}`}
            alt="community-icon"
          />
          <CommunityNameContainer>
            <CommunityNameChild>
              <h6>r/{community}</h6>
            </CommunityNameChild>
          </CommunityNameContainer>
          <JoinContainer>
            <JoinBtn
              onClick={clickHandler}
              onMouseEnter={mouseEnterHandler}
              onMouseLeave={MouseLeaveHandler}
            >
              {btnContent}
            </JoinBtn>
          </JoinContainer>
        </Item>
      </Link>
    </Container>
  <HeaderLine/>
      </>
  );
};

export default CommunityCardItem;
