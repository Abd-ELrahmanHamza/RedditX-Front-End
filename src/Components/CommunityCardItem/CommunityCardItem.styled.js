import styled from "styled-components";

export const Container = styled.div`
  a {
    margin-bottom: 0;
    border-radius: 4px 4px 0 0;
    display: flex;
    flex-direction: column;
    margin-top: -1px;
    padding: 16px;
    color: inherit;
    text-decoration: inherit;
  }
`;
export const Item = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
`;
export const Number = styled.span`
  color: ${({ theme }) => theme.color.primary};
  text-align: right;
  min-width: 20px;
  font-size: 14px;
  font-weight: 500;
  line-height: 18px;
`;
export const Caret = styled.span`
  color: rgb(70 209 96);
  font-size: 20px;
  font-weight: 400;
  vertical-align: middle;
  margin: 8px;
`;
export const Image = styled.img`
  border: none;
  background-color: #205493;
  font-size: 36px;
  height: 36px;
  max-height: 36px;
  max-width: 36px;
  overflow: hidden;
  width: 36px;
  line-height: normal;
  flex-shrink: 0;
  border-radius: 100%;
  vertical-align: middle;
`;
export const CommunityNameContainer = styled.div`
  flex-grow: 1;
  padding: 0 8px;
  overflow: hidden;
  overflow-wrap: break-word;
`;
export const CommunityNameChild = styled.div`
  display: flex;
  align-items: baseline;
  flex-direction: column;

  h6 {
    font-size: 12px;
    line-height: 16px;
    color: ${({ theme }) => theme.color.primary};
    font-weight: 700;
    max-width: 100%;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    :hover {
      text-decoration: none;
    }
  }
`;
export const JoinBtn = styled.button`
  position: relative;
  background-color: ${({ theme }) => theme.color.secondary};
  border: 1px solid transparent;
  color: ${({ theme }) => theme.background.primary};
  fill: ${({ theme }) => theme.color.secondary};
  font-family: Noto Sans, Arial, sans-serif;
  font-size: 12px;
  font-weight: 700;
  letter-spacing: unset;
  line-height: 17px;
  text-transform: unset;
  height: 24px;
  width: 56px;
  padding: 4px 16px;
  align-items: center;
  border-radius: 9999px;
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  text-align: center;
  cursor: pointer;
  flex-shrink: 0;
  outline: none;
  :hover {
    background-color: ${({ theme }) =>
      theme.background.hover_background_button_blue};
  }
`;
export const JoinContainer = styled.div`
  flex-shrink: 0;
  border: none;
`;
export const HeaderLine = styled.hr
`
  margin: 0.1rem 0;
  opacity: .1 ;
`