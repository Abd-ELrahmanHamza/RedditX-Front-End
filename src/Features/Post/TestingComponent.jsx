// Import contexts providers
import { SubmitDestinationProvider } from "Features/Post/Contexts/submitDestination";
import { CreatePostTitleProvider } from "Features/Post/Contexts/createPostTitle";
import { CreatePostFlagsProvider } from "Features/Post/Contexts/createPostFlags";
import { CreatePostAttachmentsProvider } from "Features/Post/Contexts/createPostAttachments";
import { CreatePostFlairsProvider } from "Features/Post/Contexts/createPostFlairs";
import { CreatePostTextProvider } from "Features/Post/Contexts/createPostText";
import {
  AuthProvider,
  useAuth,
} from "Features/Authentication/Contexts/Authentication";

// Import react router dom
import { BrowserRouter, Route, Routes } from "react-router-dom";

// Import themes
import darkTheme from "Theme/darkTheme";
import defaultTheme from "Theme/defaultTheme";
import lightTheme from "Theme/lightTheme";

// Import theme provider from styled components
import { ThemeProvider } from "styled-components";
import { useEffect } from "react";

const AuthTesting = ({ children }) => {
  const auth = useAuth();
  useEffect(() => {
    auth.login({
      expiresIn: 86400,
      token:
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbFR5cGUiOiJiYXJlIGVtYWlsIiwidXNlcm5hbWUiOiJ0Ml9ob21lIiwiaWF0IjoxNjY5OTI0MDEzLCJleHAiOjE2NzAzNTYwMTN9.FDa59vJi5gWy4dQVkEtgyCfm3Pp2BtHkPz0_AfNVKgY",
      username: "t2_home",
    });
  }, []);

  return (
    <BrowserRouter>
      <ThemeProvider theme={{ ...defaultTheme, ...lightTheme }}>
        <SubmitDestinationProvider>
          <CreatePostTitleProvider>
            <CreatePostFlagsProvider>
              <CreatePostAttachmentsProvider>
                <CreatePostFlairsProvider>
                  <CreatePostTextProvider>{children}</CreatePostTextProvider>
                </CreatePostFlairsProvider>
              </CreatePostAttachmentsProvider>
            </CreatePostFlagsProvider>
          </CreatePostTitleProvider>
        </SubmitDestinationProvider>
      </ThemeProvider>
    </BrowserRouter>
  );
};

const TestingComponent = ({ children }) => {
  return (
    <AuthProvider>
      <AuthTesting>{children}</AuthTesting>
    </AuthProvider>
  );
};

export default TestingComponent;
