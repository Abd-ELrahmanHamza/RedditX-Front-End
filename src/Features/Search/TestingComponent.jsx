// Import react router dom
import { BrowserRouter, Route, Routes } from "react-router-dom";
// Import themes
import darkTheme from "Theme/darkTheme";
import defaultTheme from "Theme/defaultTheme";
import lightTheme from "Theme/lightTheme";

// Import theme provider from styled components
import { ThemeProvider } from "styled-components";
import { SafeContextProvider } from "./Contexts/SafeSearchContext/Safe-context";
import { SearchContextProvider } from "./Contexts/SearchWordContext/Search-context";

const TestingComponent = ({ children }) => {
  return (
    <BrowserRouter>
      <ThemeProvider theme={{ ...defaultTheme, ...lightTheme }}>
        <SearchContextProvider>
          <SafeContextProvider>{children}</SafeContextProvider>
        </SearchContextProvider>
      </ThemeProvider>
    </BrowserRouter>
  );
};

export default TestingComponent;
