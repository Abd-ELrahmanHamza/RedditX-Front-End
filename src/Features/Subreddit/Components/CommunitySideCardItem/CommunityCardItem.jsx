import { BASE_URL } from "API/axios";
import {Container,
CommunityA,
CommunityIndex,
Arrow,
CommunityImg,
TitleParagraph,
} from "./CommunityCardItem.styled";

/**
 * Component that  shows the names of communities up-to-date.
 * @Component
 * @param {string} title - Subreddit Title
 * @param {number} index - Subreddit index among fetched data to index the list
 * @returns {Component.React}
 */
const CommunitySideCardItem = ({index, title, img}) => {
    
    return (
        <Container>
            <CommunityA to={`/subreddit/*`}>
                <CommunityIndex>{index}</CommunityIndex>
                <Arrow />
                 <CommunityImg
                    crossOrigin="anonynmous"
                    src={`${BASE_URL}/subreddits/files/${img}`}
                ></CommunityImg>
                <TitleParagraph>r/{title}</TitleParagraph> 
            </CommunityA>
        </Container>
    );
};

export default CommunitySideCardItem;