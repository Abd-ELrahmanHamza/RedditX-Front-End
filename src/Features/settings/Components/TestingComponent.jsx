import {
  AuthProvider,
  useAuth,
} from "Features/Authentication/Contexts/Authentication";

// Import react router dom
import { BrowserRouter, Route, Routes } from "react-router-dom";

// Import themes
import darkTheme from "Theme/darkTheme";
import defaultTheme from "Theme/defaultTheme";
import lightTheme from "Theme/lightTheme";

// Import theme provider from styled components
import { ThemeProvider } from "styled-components";
import { useEffect } from "react";

const AuthTesting = ({ children }) => {
  return (
    <BrowserRouter>
      <ThemeProvider theme={{ ...defaultTheme, ...lightTheme }}>
        {children}
      </ThemeProvider>
    </BrowserRouter>
  );
};

const TestingComponent = ({ children }) => {
  return (
    <AuthProvider>
      <AuthTesting>{children}</AuthTesting>
    </AuthProvider>
  );
};

export default TestingComponent;
