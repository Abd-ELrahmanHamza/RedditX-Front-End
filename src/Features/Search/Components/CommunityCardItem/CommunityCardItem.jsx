import { useState } from "react";
import { Link } from "react-router-dom";
// import logo from "../../Assets/download.jpg";
import { Container } from "./CommunityCardItem.styled";
// Import api
import axios from "API/axios";
import useFetchFunction from "Hooks/useFetchFunction";
import joinCommunity from "Features/Search/Services/joinCommunity";
/**
 * Component that contains the CommunityCardItem and manage the state of the button join.
 *
 * @Component
 * @param {String} communityIcon - The image of the CommunityCardItem
 * @param {String} communityID - The ID of the CommunityCardItem
 * @param {String} communityDescription - The Description of the CommunityCardItem
 * @param {number} membersCount - The membersCount of the CommunityCardItem
 * @param {String} communityName - The name of the CommunityCardItem
 * @param {object} isJoined - it is object indicating whether the user is joining this community
 * @returns {React.Component}
 */
const CommunityCardItem = ({
  communityIcon,
  communityID,
  communityDescription,
  communityName,
  membersCount,
  isJoined,
}) => {
  const [joinRes, errorJoin, joinLoading, fetchFunction] = useFetchFunction();
  const [isJoinedstate, setisJoined] = useState(false);

  const initialState = `${isJoined !== undefined ? "Joined" : "Join"}`;
  // the state of the buuton
  const [btnContent, setBtnContent] = useState(
    isJoined !== undefined ? "Joined" : "Join"
  );

  /**
   * it is the function that handle the state of the button when click on it.
   *
   * @param {Object} e it is an object which i can preventDefault and not redirect to the link
   */

  const clickHandler = (e) => {
    e.preventDefault();
    if (btnContent === "Join") {
      setBtnContent("Joined");
      setisJoined(true);
    } else {
      setBtnContent("Join");
      setisJoined(false);
    }
    let btnState;
    if (btnContent === "Join") {
      btnState = true;
    } else {
      btnState = false;
    }
    let dataObj = {
      action: !btnState ? "unsub" : "sub",
      sr_name: `${communityName}`,
    };
    joinCommunity(fetchFunction, dataObj);
  };
  /**
   * it is the function that handle the state of the button when mouseEnter on it.
   *
   */
  function mouseEnterHandler() {
    if (btnContent === "Joined") {
      setBtnContent("Leave");
    }
  }
  /**
   * it is the function that handle the state of the button when MouseLeave on it.
   *
   */
  function MouseLeaveHandler() {
    if (btnContent === "Leave") {
      setBtnContent("Joined");
    }
  }
  var abbreviate = require("number-abbreviate");
  console.log(communityIcon);
  return (
    <Container title="comm">
      <Link to="#">
        <div className="item">
          {communityIcon && (
            <img src={require(`../../Assets/${communityIcon}`)} alt="" />
          )}
          <div className="info">
            <div className="info2">
              <h6>r/{communityName.substring(3)}</h6>
              <p>{abbreviate(membersCount, 1)} Members</p>
            </div>
          </div>
          <div className="button">
            <button
              onClick={clickHandler}
              onMouseEnter={mouseEnterHandler}
              onMouseLeave={MouseLeaveHandler}
              className="join"
            >
              {btnContent}
            </button>
          </div>
        </div>
      </Link>
    </Container>
  );
};

export default CommunityCardItem;
