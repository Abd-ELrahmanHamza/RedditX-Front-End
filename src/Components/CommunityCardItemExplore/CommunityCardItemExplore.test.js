import { async } from "@firebase/util";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TestingComponent from "Features/Search/TestingComponent";
import CommunityCardItem from "./CommunityCardItemExplore";
// import CommunityCardItem from "./CommunityCardItem";
describe("CommunityCardItem component", () => {
  ///////////
  it('renders "join" if the button was NOT clicked', () => {
    render(
      <TestingComponent>
        <CommunityCardItem
          communityIcon="CommunityImage.png"
          communityID={"1asasddaghdgj"}
          communityDescription={
            "A subreddit dedicated to German photos and portraits from the period of 1933-1946 (dates are flexible)"
          }
          communityName={"t2_GermanWW2photos"}
          membersCount={10}
          isJoined={undefined}
        />
      </TestingComponent>
    );
    const outputElement = screen.getByRole("button");
    expect(outputElement.textContent).toBe("Join");
  });

  it('renders "Joined" if the JoinState is True', async () => {
    // Arrange
    render(
      <TestingComponent>
        <CommunityCardItem
          communityIcon="CommunityImage.png"
          communityID={"1ghhfg"}
          communityDescription={
            "A subreddit dedicated to German photos and portraits from the period of 1933-1946 (dates are flexible)"
          }
          communityName={"t2_GermanWW2photos"}
          membersCount={10}
          isJoined={true}
        />
      </TestingComponent>
    );
    // Assert
    const outputElement = screen.getByRole("button");
    expect(outputElement.textContent).toBe("Joined");
  });
  /////////////
  //////////////
  it('does not render "Join" if the button was clicked, transition from false to true', () => {
    // Arrange
    render(
      <TestingComponent>
        <CommunityCardItem
          communityIcon="CommunityImage.png"
          communityID={"fdfdfd"}
          communityDescription={
            "A subreddit dedicated to German photos and portraits from the period of 1933-1946 (dates are flexible)"
          }
          communityName={"t2_GermanWW2photos"}
          membersCount={10}
          isJoined={undefined}
        />
      </TestingComponent>
    );

    // Act
    const buttonElement = screen.queryByRole("button");
    fireEvent.click(buttonElement);

    // Assert
    const outputElement = screen.queryByText("Join");
    expect(outputElement).toBeNull();
  });

  it('does not render "Joined" if the button was clicked, transition from true to false', () => {
    // Arrange
    render(
      <TestingComponent>
        <CommunityCardItem
          communityIcon="CommunityImage.png"
          communityID={"fgfg"}
          communityDescription={
            "A subreddit dedicated to German photos and portraits from the period of 1933-1946 (dates are flexible)"
          }
          communityName={"t2_GermanWW2photos"}
          membersCount={10}
          isJoined={true}
        />
      </TestingComponent>
    );

    // Act
    const buttonElement = screen.queryByRole("button");
    fireEvent.click(buttonElement);

    // Assert
    const outputElement = screen.queryByText("Joined");
    expect(outputElement).toBeNull();
  });
  it("renders the community in correct way in Card", async () => {
    render(
      <TestingComponent>
        <CommunityCardItem
          communityIcon="CommunityImage.png"
          communityID={"fgfg"}
          communityDescription={
            "A subreddit dedicated to German photos and portraits from the period of 1933-1946 (dates are flexible)"
          }
          communityName={"t2_GermanWW2photos"}
          membersCount={10}
          isJoined={true}
        />
      </TestingComponent>
    );
    const outputname = screen.getByText("r/GermanWW2photos");
    expect(outputname).toBeInTheDocument();
    // const outputmembers = screen.getByText("{membersCount}m Members");
    // expect(outputmembers).toBeInTheDocument();
  });
});
//test
