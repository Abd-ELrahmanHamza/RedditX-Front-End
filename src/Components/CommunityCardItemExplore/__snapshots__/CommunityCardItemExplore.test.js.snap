// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Community CardI tem this is a test for Community Card Item 1`] = `
<styled.div
  title="comm"
>
  <Link
    to="/subreddit/t5_imagepro"
  >
    <div
      className="item"
    >
      <img
        alt=""
        crossOrigin="anonynmous"
        src="https://api.redditswe22.tech/subreddits/files/CommunityImage.png"
      />
      <div
        className="info"
      >
        <div
          className="info2"
        >
          <div>
            r/
            manWW2photos
          </div>
          <p>
            10
             Members
          </p>
        </div>
      </div>
      <div
        className="button"
      >
        <button
          className="join"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Join
        </button>
      </div>
    </div>
  </Link>
</styled.div>
`;
